feat: add HTTPClient implementation using curl as a module

jonbonazza opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->
This PR uses libcurl in multi mode (curl_multi) to implement the HTTPClient interface. It cleans up the implementations significantly and should reduce maintenance burden while giving us easier access to more features in the end.

fixes https://github.com/godotengine/godot-proposals/issues/2903

## Scope

The goal of this PR is only to create a curl-based implementation of `HTTPClient` that conforms to the existing interface and nothing more.

The net team has discussed several improvements to the `HTTPClient` interface in general that may be implemented in future PRs, but they remain outside the scope of this one.

## Remaining Work

The current state of the PR is *DRAFT*. It still needs more testing and there are still a few remaining pieces left.

- [x] Support IPv6 DNS resolution using Godot's resolver APIs. I was unable to get this to work, but I consider it essential for this PR to be merged. IPv4 works fine though.
- [x] See if there's a way to make DNS resolution more performant by using the async apis. I'm not familiar enough with Godot's DNS APIs to know how best to use them, and the documentation for them is quite lacking. May want to just leave this up to a future improvement with its own PR.
- [x] Implement `set_connection` and `get_connection`. The way libcurl works doesn't lend itself well to the `StreamPeer` interface, so without some changes to that API, there's not a good way to do t his. If we _do_ decide to use `NetSocket` as the socket implementation (see above item) then we get this for free, so that's one pro to that approach. That said, I don't see these functions used anywhere in the Godot codebase, so I'd question whether they are even needed to begin with.
- [x] Get libcurl building on all platforms. Currently, I only have it working on Windows. I know @Faless has gotten it working on Linux, but I'm not sure what was necessary there. This might take some effort. We should also take the time to decide what features to enable/disable to tune binary size and also add a flag for using the system libcurl if available.


If I've missed anything in either the implementation itself, or the TODO list above, please don't hesitate to bring it up so we can discuss and potentially address it.

## Regarding Unit Tests

While we should certainly strive for as much unit tests coverage as possible, I didn't see any existing tests for the http client, so I've decided to leave that for an exercise for another day (and potentially another person). 

I want to stress that my decision not to include unit tests is not to the discredit of the idea of unit tests in general; quite the contrary, in fact. I would like to see a day where Godot's unit test coverage nears 100% just as much as anyone. However the best way to test http code is by running it against a real HTTP server, and to my knowledge, Godot doesn't currently have the facilities to do that and I'm not comfortable convoluting this particular PR by adding those facilities.

## A Note About The Commits

There's quite a lot of commits in this one and none of them are particularly clean. I'm a bit embarrassed about that, but it's mostly due to the fact I was learning so many new technologies along the way and ended up doing a lot of things wrong in a lot of different places in odd orders.

I suggest when merging this PR to squash all of the commits in order to keep things as clean as possible in master.

Calinou commented:
> Currently, we only resolve one address for a domain. Godot's resolver APIs support resolving multiple

Wasn't this implemented by https://github.com/godotengine/godot/pull/49026?

Faless commented:
Few things also mentioned in the chat:
- **Only fetch the body during `read_response_body_chunk`**. Not during poll :exclamation: 
- For IP resolution, see:
  https://github.com/godotengine/godot/blob/0caaaf4018feddb1d3e6e75870df282715efd276/core/io/http_client_tcp.cpp#L81-L83
  https://github.com/godotengine/godot/blob/0caaaf4018feddb1d3e6e75870df282715efd276/core/io/http_client_tcp.cpp#L250-L255


> * [blocking] I'm not even really sure how this is expected to be used. Is it even needed?

When you want to handle the request in a thread.

> * Implement `set_connection` and `get_connection`.

If we don't use our NetSocket implementation, just have them do nothing/return null and spit an error. Like JavaScript does:

https://github.com/godotengine/godot/blob/0caaaf4018feddb1d3e6e75870df282715efd276/platform/javascript/http_client_javascript.cpp#L74-L76

> * Docs... I dread this one.

Well, be happy, you won't have to write any new doc, given you are creating a new implementation of an exposed class, not a new exposed class (so HTTPClient is already fully documented).
You can of course try to improved the documentation, if you like :).

> * Get libcurl building on all platforms.

Yeah, this is going to be quite annoying, on Linux I only got it to work with system-wide mbedtls (not with built-in) and using @fire `curl_config.h`.

Faless commented:
> While we should certainly strive for as much unit tests coverage as possible, I didn't see any existing tests for the http client, so I've decided to leave that for an exercise for another day (and potentially another person).

I had testing for HTTPClient on a separate repository like the one you are proposing (those are not unit tests):
https://github.com/Faless/gd-tests/

We can use those to verify the correct behaviour, the main problem was that websites would, from time to time slightly change their layout, so of course tests would unexpectedly break.

Faless commented:
Size comparison (`target=release, builtin_mbedtls=no`):
Curl: `53436 KiB`
NoCurl: `53176 KiB`
Diff: `~260 KiB`

jonbonazza commented:
> > Currently, we only resolve one address for a domain. Godot's resolver APIs support resolving multiple
> 
> Wasn't this implemented by #49026?
>

Yes, the engine supports it. This code, however does not. I am also unsure whether it's possible to do in an async manner. It's also worth mentioning that the current HTTPClient implementation does not seem to do multi address resolution, so Im gonna leave it to be outside the scope of this PR for now.

jonbonazza commented:
Okay I've trimmed down the TODO list a bit.

> Only fetch the body during read_response_body_chunk. Not during poll ❗

Fixed this.

Also, @Faless I've updated the code to use the async resolver polling. It's basically the same code as the existing HTTPClientTCP you referenced, so nothing too crazy there.

Still can't get IPv6 to work and still not quite groking the blocking mode stuff. I've left a couple messages in the dev chat.

jonbonazza commented:
This is ready for review.

I've tested this on Windows using the attached project, however it should probably be tested on linux, os, anxdroid, and ios as well for posterity.

[curl_test.zip](https://github.com/godotengine/godot/files/7071993/curl_test.zip)

Faless commented on modules/mbedtls_curl/http_client_curl.h:
```diff
@@ -0,0 +1,126 @@
+	bool ssl = false;
+	bool verify_host = false;
+	bool blocking_mode = false;
+	int read_chunk_size = 0;
```
```suggestion
	int read_chunk_size = 65536;
```

Faless commented:
Would be great to squash, leaving 2 commits, one adding the third-party library the other adding the module itself.

Faless commented on thirdparty/README.md:
```diff
@@ -35,6 +35,19 @@ Files extracted from upstream source:
+File extracted from a recent Fedora install:
+/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
+(It can't be extracted directly from the package,
+as it's generated on the user's system.)
```
I think this is from a rebase issue?

fire replied:
I resolved this issue because I think the original bundle is required for the fallback.

Faless commented on thirdparty/README.md:
```diff
@@ -35,6 +35,19 @@ Files extracted from upstream source:
+
+Files extracted from upstream source:
+
+- all .c and .h files in the src, include, and lib directories (and their descendants).
```
we should mention our custom config too

fire replied:
Mentioned the derived config.

fire commented:
@jonbonazza Will work on rebasing into two prs and resolving the reviews.

fire committed:
Add libcurl module.

Co-authored-by: Jon Bonazza <jonbonazza@gmail.com>
Co-authored-by: Fabio Alessandrelli <fabio.alessandrelli@gmail.com>

fire commented:
Historical note: The module name `mbedtls_curl` was done to force this module to build after `mbedtls`.

fire commented:
```bash
In file included from thirdparty/curl/lib/vtls/mbedtls.c:59:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
Compiling ==> thirdparty/curl/lib/bufref.c
1 error generated.
scons: *** [thirdparty/curl/lib/vtls/mbedtls.windows.opt.tools.64.o] Error 1
In file included from thirdparty/curl/lib/altsvc.c:35:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
1 error generated.
scons: *** [thirdparty/curl/lib/altsvc.windows.opt.tools.64.o] Error 1
In file included from thirdparty/curl/lib/asyn-thread.c:67:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
1 error generated.
scons: *** [thirdparty/curl/lib/asyn-thread.windows.opt.tools.64.o] Error 1
In file included from thirdparty/curl/lib/vtls/vtls.c:55:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
1 error generated.
scons: *** [thirdparty/curl/lib/vtls/vtls.windows.opt.tools.64.o] Error 1
```

Using mingw llvm.

>  /bin/bash -c "PATH=/opt/llvm-mingw/bin:$PATH scons werror=no platform=windows target=release_debug -j`nproc` use_lto=no deprecated=no use_mingw=yes use_llvm=yes use_lld=yes use_thinlto=yes LINKFLAGS=-Wl,-pdb= CCFLAGS='-g -gcodeview' debug_symbols=no custom_modules=../godot_custom_modules"

fire commented:
Discussed with Lyuma this might be a VARIADIC define missing. Can get a rebase in near future.

fire commented:
@Faless @jonbonazza Can you verify the pr? I think it's ready to be merged.

Faless commented on modules/mbedtls_curl/http_client_curl.h:
```diff
@@ -0,0 +1,128 @@
+#include <curl/curl.h>
+#include <stdio.h>
+
+struct RequestContext {
+	RequestContext(){};
+	~RequestContext();
+
+	List<String> *response_headers = nullptr;
+	int *response_code = nullptr;
+	RingBuffer<uint8_t> *read_buffer = nullptr;
+	curl_slist *header_list = nullptr;
+	int *body_size = nullptr;
+	HTTPClient::Status *status = nullptr;
+	Vector<PackedByteArray> *response_chunks = nullptr;
+	bool *has_response = nullptr;
+	bool *chunked = nullptr;
+	bool *keep_alive = nullptr;
+};
```
Is there any reason to use this `RequestContext` with all pointers instead of just passing the `HTTPClientCurl` itself to (i.e. `this`) to the callbacks?
If we go for the `RequestContext` approach (which could be also good for future-proofing), those members should not be pointers, but the actual types.
We can add a `RequestContext current_context` member to `HTTPClientCurl` and pass the pointer to that to the callbacks (and add a `RequestContext::clear` for clearing it).

jonbonazza replied:
Okay @Faless , I've updated the PR to remove the `RequestContext` struct entirely. It not stores the request state in the HttpClient object itself. I remember now that the reason I implemented the RequestContext was because I wasn't sure whether or not there could be multiple requests for the same client in flight at once. It doesn't seem like there can be given the way the interface is designed.

Faless commented on modules/mbedtls_curl/http_client_curl.cpp:
```diff
@@ -0,0 +1,440 @@
+		h = "[" + h + "]";
+	}
+	CURL *eh = curl_easy_init();
+	curl_easy_setopt(eh, CURLOPT_URL, (scheme + h + ":" + String::num_int64(port) + url).ascii().get_data());
```
Here, we don't want to set the ip address as the `CURLOPT_URL` but the actual hostname (this is what will be used for the `Host: hostname:port` header).

jonbonazza replied:
Okay this is fixed. And it also makes `http://www.example.com` finally work as well!

Faless commented on modules/mbedtls_curl/http_client_curl.cpp:
```diff
@@ -0,0 +1,440 @@
+Error HTTPClientCurl::_poll_curl() {
+	CURLMcode rc = curl_multi_perform(curl, &still_running);
+	if (still_running) {
+		rc = curl_multi_wait(curl, nullptr, 0, 1000, nullptr);
```
This is blocking (`1000` is the timeout). Should probably be:
```
curl_multi_wait(curl, nullptr, 0, 0, nullptr)
```

And I think we can use this to implement the blocking mode.

jonbonazza replied:
@Faless okay this is done. It seems to work as expected.

Faless commented on modules/mbedtls_curl/http_client_curl.cpp:
```diff
@@ -0,0 +1,440 @@
+		}
+	}
+	if (response_chunks.is_empty()) {
+		status = keep_alive ? STATUS_CONNECTED : STATUS_DISCONNECTED;
```
So the big problem with this is that between one `_poll_curl()` and the other there might not have been any new data received, which will be interpreted as the request being over.

jonbonazza replied:
Alright @Faless I'm not terribly happy with my solution to this, but it's the best I could think of after several hours of racking my brain. I left a comment in the code describing the problem and the up- and downsides to my solution.

Faless reviewed (changes requested):
This looks very promising, I did spend a few hours hacking around it, testing various URLs, downloads, etc.
The biggest problem I have now is that I haven't found a reliable way to detect when curl has finished parsing the response (see my comment on the `STATUS_CONNECTED/DISCONNECTED` change).

We need to be able to do that for it to work properly in all cases (chunked/keep-alive/read-till-eof).

Faless commented on modules/mbedtls_curl/register_types.cpp:
```diff
@@ -0,0 +1,52 @@
+		curl_ok = true;
+	}
+
+	GDREGISTER_CLASS(HTTPClientCurl);
```
Forgot about this,
This does not need be registered, here, you would just call `HTTPClientCurl::make_default()` so API compatibility with and without the module is preserved.

jonbonazza replied:
Done.

jonbonazza commented:
Thanks for the review @Faless and thank you soooooo much for the assistance with the build stuff @fire ! I unfortunately probably won't be able to get to addressing the comments tonight like I wanted to, but I'm confident that I'll be able to knock them out this week. 🤞 I left some follow up comments/questions on some of the concerns.

fire commented on modules/mbedtls_curl/http_client_curl.cpp:
```diff
@@ -0,0 +1,471 @@
+		// the connection fails. Simply put, it means it has to have the same
+		// name in the certificate as is in the URL you operate against.
+		// @see https://curl.se/libcurl/c/CURLOPT_SSL_VERIFYHOST.html
+		curl_easy_setopt(eh, CURLOPT_SSL_VERIFYHOST, 0L);
```
Was this supposed to be 2?

fire replied:
The diff was confusing. This is correct to not verify host. 

Can you rebase to like 2-3 commits?

jonbonazza replied:
Yep. Ive actually already done so locally. I was just waiting for CI to finish before I pushed. Ill do so shortly.

jonbonazza commented:
@fire I've reduced the commits down to 2.

Faless commented:
It doesn't so handle HEAD requests correctly:

```
@onready
var req = $HTTPRequest

# Called when the node enters the scene tree for the first time.
func _ready():
        req.connect("request_completed", Callable(self, "_completed"))
        # Stalls instead of completing after the header like the current implementation does.
        req.request("https://google.com/", [], false, HTTPClient.METHOD_HEAD)

```

[proj_httprequest_head.zip](https://github.com/godotengine/godot/files/7857018/proj_httprequest_head.zip)


Please refer to https://github.com/Faless/gd-tests/ (quite outdated) for some test URL.

Test at least:
- [ ] GET/HEAD/POST both SSL and not, both with known content length and without.
- [ ] Chunked/Content-Length keep-alive.
- [ ] Read-till-EOF.

Would be better to also test:
- [ ] PUT/DELETE (very common in rest APIs)
- [ ] OPTIONS (not sure, but should be supported by the current implementation, for CORS pre-flight).

jonbonazza commented:
@Faless okay, I've tested every HTTP method I could think of and both chunked and not chunked for the ones that made sense. I also went through all the tests in your test project and made sure my responses matched what the tests expected. I also added all of the default headers that the old http client was adding.

One thing is I did screw up the git history so I ended up just squashing everything down.

jonbonazza commented:
FWIW, a suuuper handy tool for doing these tests was https://httpbin.org

mhilbrunner commented:
This is super not useful but I just want to chime in and say that its awesome how consistently you've been working away on this. Thanks for the efforts!

jonbonazza commented:
More bugs were founds. Squashing. 🐞

jonbonazza commented:
Ooookay. Squashed all the bugs I found. @fire the issue you found should also be resolved.

fire commented:
I can still see a bug. Having trouble reducing a test case.

int-72h commented:
This feature looks great (as at the moment I'm having to use a GDNative extension), will it be merged anytime soon?

andy5995 commented:
> This feature looks great (as at the moment I'm having to use a GDNative extension), will it be merged anytime soon?

Looks like it has merge conflicts at the moment.
